Add a REST API output caching engine

Konamiman opened the pull request:
_**NOTE:** This pull request has been divided in two: https://github.com/woocommerce/woocommerce/pull/61600 and https://github.com/woocommerce/woocommerce/pull/61601_

### Changes proposed in this Pull Request:

This pull requests adds a transient-based caching engine for the output of REST API endpoints, in the form of a `RestApiCache` trait that can be added to any controller. Features:

- By default the engine activates only if an external object cache is in place (if `wp_using_ext_object_cache()` returns true). See note below.
- Caches the output of any endpoint whose callback is wrapped into a `$this->with_cache` call.
- Caches response headers too, excluding a fixed set (including e.g. `Date`) and a customizable (per-controller and per-endpoint) set.
- Requests are uniquely identified by the combination of the request route and the query string contents, and by default (configurable per-controller and per-endpoint) also by the current user id. Controllers can customize this.
- Entity version tracking implemented for invalidation of cached responses involving modified entries. Entity versions are stored in transients too, using the new `EntityVersionsCache` class.
- Extracts the list of relevant entity ids from the response (for entity version tracking) assuming an `id` field is present, this behavior can be customized per controller and per endpoint.
- Takes in account filters that modifiy the response before it's sent: if the hookings to these filters change, cached responses get invalidated. The list of relevant filters can be specified per controller and per endpoint.
- Configurable lifetime for cached responses, defaulting to one hour. This behavior can be customized per controller and per endpoint.
- Configurable (via filter) lifetime for cached entity versions, defaults to one hour. 
- Implements an `X-WC-Cache` response header, with values "HIT", "MISS" or "SKIP".
- Caching can be disabled for a given request by adding `_skip_caching=true|1` to the query string.
- Caching can be disabled per request using a filter (`woocommerce_rest_api_enable_response_caching`), a `must_cache` method in the controller, or endpoint caching configuration.

This lays the foundation for implementing actual caching in existing REST API controllers, and for ETag-based caching of REST API responses later on.

#### Note about the usage of transients

This implementation uses the WordPress transients API to cache REST API endpoints outputs and entity versions. By default WordPress uses the options table to store transients, and in that case using transients as a heavy-duty cache is not a good idea; but if an external object cache is in place (which is by default mandatory for this caching engine to activate), then that cache is used instead to store the transients.

That's the reason why the transients API is used instead of the `wp_cache_*` functions here: transients will be stored in the external cache anyway if it's active, and if not, the caching engine can be activated anyway (for development or testing purposes) via a dedicated filter.


### How to test the changes in this Pull Request:

Note: these testing instructions cover basic scenarios but there are more per-controller and per-endpoint customizations available, these are all covered by unit tests but feel free to take a look at the docblock and the code of `RestApiCache` and tinker with additional method overrides, filters and configurations passed to `with_cache`.

Prerrequisite: either install a persistent object cache in your WordPress instance, or add the following snippet: `add_filter('woocommerce_enable_entity_versions_cache', '__return_true');` - without doing one of these, the cache engine won't activate and you won't see any changes in behavior compared to `trunk`.

1. Add the following snippet to your site, this will create a simple "things" controller for testing the caching trait:

```php
add_action( 'rest_api_init', function() {
    $controller = new class() extends WP_REST_Controller {
        use Automattic\WooCommerce\Internal\Traits\RestApiCache;

        public function __construct() {
            $this->initialize_rest_api_cache();
        }

        protected function get_default_entity_type(): ?string {
            return 'thing';
        }

        public function register_routes() {
            register_rest_route(
                'wc',
                '/things',
                array(
                    'methods'             => WP_REST_Server::READABLE,
                    'callback'            => $this->with_cache(
                        array( $this, 'get_items' ),
                        array(
                            'cache_ttl'       => 60, // 1 minute
                            'relevant_hooks'  => array( 'woocommerce_modify_things' ),
                            'exclude_headers' => array( 'X-Exclude-Me' )
                        )
                    ),
                    'permission_callback' => '__return_true',
                )
            );
        }

        public function get_items( $request ) {
            // 1 second delay just for fun
            sleep( 1 );

            $data = array(
                array( 'id' => 1, 'name' => 'Thing 1' ),
                array( 'id' => 2, 'name' => 'Thing 2' ),
            );

            $data = apply_filters( 'woocommerce_modify_things', $data, $request );

            $response = rest_ensure_response( $data );
            $response->header( 'X-Custom-Header', 'CustomValue-' . time() );
            $response->header( 'X-Exclude-Me', 'Ok' );
            $response->header( 'Cache-Control', 'No, thank you.' );
            return $response;
        }
    };

    $controller->register_routes();
} );
```

2. Run the following command:

```
time curl -i http://localhost/wp-json/wc/things
```

The time should be slightly over 1s, and the response should contain a `X-WC-Cache: MISS` header. Also you should see the `X-Custom-Header`, `X-Exclude-Me` a `Cache-Control` headers in the response.

3. Wait a few seconds and repeat the command. Now the response should come much faster, and the response should contain a `X-WC-Cache: HIT` header. You should see the `X-Custom-Header` header with the same value in the response, and you should *not* see a `X-Exclude-Me` nor a `Cache-Control` header.

4. Wait one minute and repeat the request twice, you should get again a MISS followed by a HIT. (Now you can increase the value of `'cache_ttl'` to ease testing if you want).

4. Repeat the request twice adding `?foo=bar` to the query string, you should get again a MISS followed by a HIT.

5. Add the following additional snippet:

```php
add_filter( 'woocommerce_modify_things', function( $data, $request ) {
    $data[0]['modified'] = true;
    return $data;
}, 10, 2 );
```

and again repeat the request twice, you should get again a MISS followed by a HIT.

6. Run the following: `wp eval 'wc_get_container()->get( Automattic\WooCommerce\Internal\Caches\EntityVersionsCache::class )->modify_entity_version( "thing", 1 );'` - repeating the request should again result in a MISS followed by a HIT.

7 Add `?_skip_cache=true` to the query string, now the response should always be slow and include `X-WC-Cache: SKIP`.

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

Konamiman committed:
Add the REST API caching engine and apply it to the products controller.

Konamiman committed:
Add changelog file

Konamiman committed:
Merge branch 'trunk' into add-rest-api-cache-engine

Konamiman committed:
Check if the current route actually belongs to the controller handling the request

Konamiman committed:
Fix: non-GET requests not being actually skipped

Konamiman committed:
Fix incorrect call to parent extract_entity_ids

Konamiman committed:
Properly handle changes in variations

Konamiman committed:
Various changes in how the mutable entities versioning is handled.

- Added the EntityVersionsCache class, which is responsible for storing
  and managing the versions of mutable entities.
- Product change events moved to the EntityVersionsCache class.
- The entity versions cache, and thus the REST API output caching,
  is now enabled only if there's an external persistent object cache
  configured. Force-enable is allowed anyway via filter.

Konamiman committed:
Merge branch 'trunk' into add-rest-api-cache-engine

Konamiman committed:
Fix linting issues

Konamiman committed:
Merge branch 'trunk' into add-rest-api-cache-engine

Konamiman committed:
Merge branch 'trunk' into add-rest-api-cache-engine

joshuatf commented on plugins/woocommerce/src/Internal/Traits/RestApiCache.php:
```diff
@@ -0,0 +1,484 @@
+	 *
+	 * @return string|null Entity type (e.g., 'product', 'variation'), or null if no default.
+	 */
+	protected function get_default_entity_type(): ?string {
```
It may not be very RESTful, but are there scenarios we need to consider where multiple entity types could exist?

Konamiman replied:
This is just a default. Controllers can override `get_request_uid_info` to provide a different type for a given request.

If you mean that a given request might return entities of different types: maybe, but for this first iteration of the engine I wanted to keep things simple. After all, the entity type is just a clue to generate cache keys so and anything will do, e.g. `orders-and-refunds` if needed. We can revisit that later if needed.

joshuatf commented on plugins/woocommerce/src/Internal/Traits/RestApiCache.php:
```diff
@@ -0,0 +1,484 @@
+
+		$uid_info = array(
+			'request_hash' => md5( $request->get_route() . '-' . wp_json_encode( $request->get_query_params() ) ),
+			'entity_type'  => $this->get_default_entity_type(),
```
Is the entity type absolutely needed?  Could the responsibility of invalidating being offset to an interface that needs to be implemented by the controller?

Konamiman replied:
The entity type is needed by `EntityVersionsCache` in order to know which entity is to be handled (entities are uniquely identified by type and id). Controllers are still in charge as they can override `get_default_entity_type` and `get_request_uid_info`.

Invalidation of cached entities (via modification of the entity version) can be done in multiple ways: controllers can listen to changes in the entities they manage (e.g. product modified hooks) and invalidate by themselves, or `EntityVersionsCache` can be used directly by the code that modifies these entities. I think this covers all the possible use cases.

joshuatf commented on plugins/woocommerce/src/Internal/Traits/RestApiCache.php:
```diff
@@ -0,0 +1,484 @@
+			return $result;
+		}
+
+		$uid_info = $this->get_request_uid_info( $request );
```
I think it would be better to look into wrapping each request or adding callbacks to the controller's endpoints instead of adding this logic to all endpoints (even those outside of Woo).

Konamiman replied:
That's a good point, I'll work on changing that.

joshuatf commented on plugins/woocommerce/src/Internal/Traits/RestApiCache.php:
```diff
@@ -0,0 +1,484 @@
+	 * @return bool True on success, false on failure.
+	 */
+	protected function set_cached_data( string $key, $value, int $ttl ): bool {
+		return set_transient( $key, $value, $ttl );
```
We should use the WP object cache instead of transients.

Konamiman replied:
Yes, we discussed that, but the thing is that:

* By default, this engine will only activate if an object cache is enabled in the site. And when that's the case, the transients functions will actually use the WP object cache.
* By leaving the usage of transients functions in place, we leave the option (even if we discourage it) to use the caching engine with the default database-backed transients engine (via the `woocommerce_enable_entity_versions_cache` filter). This is useful for testing, troubleshooting, staging, or even maybe very small production scenarios.

joshuatf commented on plugins/woocommerce/src/Internal/Caches/EntityVersionsCache.php:
```diff
@@ -0,0 +1,201 @@
+	/**
+	 * Creates a new instance of the class.
+	 */
+	public function __construct() {
```
Currently these all look very product specific and mismatch the class name which appears generic (`EntityVersionsCache`).  I think the individual controller should have an interface for declaring when cache should be invalidated and be responsible for implementing these hooks that determine invalidation.

If these are commonly used by a lot of controllers, then we can still move that into a reusable class or trait, but I think this logic doesn't belong to all entity caches.

Konamiman replied:
Good point, and as you suggested me earlier, it's better to leave this pull request as infrastructure-only and move all the actual caching for product endpoints to a separate one.

joshuatf reviewed (commented):
I like the general direction of this.  I think we need to start with a simple abstraction for cache and then a simple interface for invalidating that cache by individual endpoints.

That will make reviewing and agreeing on what the API looks like much faster.

Konamiman committed:
Modify EntityVersionsCache:

- Handle the cache in dedicated protected methods
- Add woocommerce_entity_version_cached and
  woocommerce_entity_version_cached actions

Konamiman committed:
Re-engineer RestApiCache:

Instead of hooking into the pre- and post-dispatch hooks,
now each endpoint must explicitly opt-in to caching.
This is done by wrapping the endpoint callback into a "with_cache"
method that additionally accepts configuration parameters.

Konamiman committed:
Add unit tests for RestApiCache and EntityVersionsCache

Konamiman committed:
Remove all product-specific code from controllers and EntityVersionsCache

Konamiman committed:
Don't add a cache skip header if EntityVersionsCache is disabled

Konamiman committed:
Fixes to documentation

Konamiman committed:
Fix linting errors

Konamiman committed:
Fix bug in generate_hooks_hash() method

The method was incorrectly iterating over WP_Hook objects directly
instead of accessing their callbacks property. This caused the
hook-based cache invalidation to fail.

Konamiman commented on plugins/woocommerce/src/Internal/Traits/RestApiCache.php:
```diff
@@ -0,0 +1,595 @@
+		);
+
+		return md5( wp_json_encode( $cache_hash_data ) );
+	}
```
Fixed in [`72e98ad` (#61414)](https://github.com/woocommerce/woocommerce/pull/61414/commits/72e98ad91088345bfc6fcf9ab18d9d20f1b05fd4)

Konamiman committed:
Small fixes (typos et al)

Konamiman committed:
Optimization: run build_cache_config only once per request

Konamiman commented on plugins/woocommerce/src/Internal/Traits/RestApiCache.php:
```diff
@@ -0,0 +1,595 @@
+		if ( is_null( $this->entity_versions_cache ) ) {
+			return call_user_func( $callback, $request );
+		}
+
```
That's on purpose: if the entity versions cache is disabled, the endpoint should act exactly as if the output caching engine didn't exist.

Konamiman committed:
Add optional 'vary by user' for cached responses

Konamiman commented on plugins/woocommerce/src/Internal/Traits/RestApiCache.php:
```diff
@@ -0,0 +1,595 @@
+	/**
+	 * Check if caching should be used for a particular incoming request.
+	 *
+	 * @param WP_REST_Request $request The request object.
+	 * @param array           $config  Caching configuration.
+	 * @return bool True if caching should be used, false otherwise.
+	 */
+	private function should_use_cache_for_request( WP_REST_Request $request, array $config ): bool {
+		// Check for explicit skip parameter.
+		$skip_cache   = $request->get_param( '_skip_cache' );
+		$should_cache = ! ( 'true' === $skip_cache || '1' === $skip_cache );
+
+		// Check custom must_cache callback.
+		if ( $should_cache ) {
+			$must_cache_fn = $config['must_cache'] ?? array( $this, 'must_cache' );
+			$should_cache  = call_user_func( $must_cache_fn, $request );
+		}
+
+		/**
+		 * Filter whether to enable response caching for a given REST API controller.
+		 *
+		 * Based on the _skip_cache parameter and must_cache callback evaluation,
+		 * the first parameter indicates whether caching should be enabled.
+		 * This filter can override that decision.
+		 *
+		 * @since 10.4.0
+		 *
+		 * @param bool            $enable_caching Whether to enable response caching (result of !_skip_cache && must_cache evaluation).
+		 * @param object          $controller     The controller instance.
+		 * @param WP_REST_Request $request        The request object.
+		 * @return bool True to enable response caching, false to disable.
+		 */
+		return apply_filters(
+			'woocommerce_rest_api_enable_response_caching',
+			$should_cache,
+			$this,
+			$request
+		);
+	}
```
vary-by-user scenario addressed in [`d26e655` (#61414)](https://github.com/woocommerce/woocommerce/pull/61414/commits/d26e6556d66b5a7bb7925ca4d0ea19360fd96350), non-GET responses aren't an issue because the caching mechanism is opt-in per endpoint.

Konamiman replied:
I have pushed a modification for vary_by_user to be true by default. That said, there's no need to worry about what extension developers will do: the `RestApiCache` trait is in the `Internal` namespace, thus it's intended to be used in WooCommerce core only. If at some point we decide to make it public, we'll consider adding additional safety measures.

Konamiman committed:
Improve RestApiCache: add endpoint IDs and header caching, make vary-by-user true by default

- Add optional endpoint_id parameter to all customization methods
  - Include endpoint_id in cache config and pass to all filters/actions
- Ensure all filters receive controller instance ($this)
- Implement response headers caching with configurable exclusions:
  - Always exclude: X-WC-Cache, Set-Cookie, Date, Expires,
    Last-Modified, Age, ETag, Cache-Control, Pragma
  - Support custom exclusions via controller method or endpoint config
- Add woocommerce_rest_api_cache_key_info filter
- Increase cached entity versions default ttl to one day

Konamiman committed:
Privatize some protected methods

Konamiman commented on plugins/woocommerce/src/Internal/Traits/RestApiCache.php:
```diff
@@ -0,0 +1,595 @@
+
+		$response->header( 'X-WC-Cache', $cached ? 'MISS' : 'SKIP' );
+		return $response;
+	}
```
I have pushed a change for caching response headers, with a combination of fixed and customizable exclusion lists.

Konamiman committed:
Improvement: treat query strings as unique regardless of parameter order

Konamiman committed:
Fix unstable hooks hash causing spurious cache invalidations

  Raw callbacks (closures, objects, arrays) serialized inconsistently,
  causing false cache MISSes. Now normalize all callbacks to stable
  strings before hashing to ensure consistent hash generation.

Konamiman commented:
This pull request has been divided in two: https://github.com/woocommerce/woocommerce/pull/61600 and https://github.com/woocommerce/woocommerce/pull/61601
